core/platform/lf_zephyr_support.c

```diff
@@ -0,0 +1,527 @@
+    int res = k_mutex_lock(mutex, K_FOREVER);
+    return res;
+}
+
+/** 
+ * Unlock a mutex.
+ * 
+ * @return 0 on success, platform-specific error number otherwise.
+ */
+int lf_mutex_unlock(lf_mutex_t* mutex) {
+    int res = k_mutex_unlock(mutex);
+    return res;
+}
+
+
+/** 
+ * Initialize a conditional variable.
+ * 
+ * @return 0 on success, platform-specific error number otherwise.
+ */
+int lf_cond_init(lf_cond_t* cond) {
+    return k_condvar_init(cond);
+}
+
+/** 
+ * Wake up all threads waiting for condition variable cond.
+ * 
+ * @return 0 on success, platform-specific error number otherwise.
+ */
+int lf_cond_broadcast(lf_cond_t* cond) {
+    k_condvar_broadcast(cond);
+    return 0;
+}
+
+/** 
+ * Wake up one thread waiting for condition variable cond.
+ * 
+ * @return 0 on success, platform-specific error number otherwise.
+ */
+int lf_cond_signal(lf_cond_t* cond) {
+    return k_condvar_signal(cond);
+}
+
+/** 
+ * Wait for condition variable "cond" to be signaled or broadcast.
+ * "mutex" is assumed to be locked before.
+ * 
+ * @return 0 on success, platform-specific error number otherwise.
+ */
+int lf_cond_wait(lf_cond_t* cond, lf_mutex_t* mutex) {
+    return k_condvar_wait(cond, mutex, K_FOREVER);
+}
+
+/** 
+ * Block current thread on the condition variable until condition variable
+ * pointed by "cond" is signaled or time pointed by "absolute_time_ns" in
+ * nanoseconds is reached.
+ * 
+ * @return 0 on success, LF_TIMEOUT on timeout, and platform-specific error
+ *  number otherwise.
+ */
+int lf_cond_timedwait(lf_cond_t* cond, lf_mutex_t* mutex, instant_t absolute_time_ns) {
+    
+    // lf_sleep_until(absolute_time_ns);
+    // instant_t now;
+    // lf_clock_gettime(&now);
+    // return LF_TIMEOUT;
+    instant_t now;
+    lf_clock_gettime(&now);
+    interval_t sleep_duration_ns = absolute_time_ns - now;
+    k_timeout_t timeout = K_NSEC(sleep_duration_ns);
+    int res = k_condvar_wait(cond, mutex, timeout);
+    if (res == 0) {
+        return 0;
+    } else {
+        return LF_TIMEOUT;
+    }
+}
+
+// Atomics
+//  Implemented by just entering a critical section and doing the arithmetic
+//  FIXME: We are now restricted to atomic integer operations
+int _zephyr_atomic_fetch_add(int *ptr, int value) {
+    lf_critical_section_enter();
+    int res = *ptr;
+    *ptr += value;
+    lf_critical_section_exit();
+    return res;
+}
+
+int _zephyr_atomic_add_fetch(int *ptr, int value) {
+    lf_critical_section_enter();
+    int res = *ptr + value;
+    *ptr = res;
+    lf_critical_section_exit();
+    return res;
+}
+
+// FIXME: Are you sure that it should return bool?
+bool _zephyr_bool_compare_and_swap(bool *ptr, bool value, bool newval) {
+    lf_critical_section_enter();
+    bool res = false;
+    if (*ptr == value) {
+        *ptr = newval;
+        res = true;
+    }
+    lf_critical_section_exit();
+    return res;
+}
+
+// FIXME: Are you sure that it should return bool?
```

Comment:
When these two functions were added to our platform files, they were based directly on [GCC's atomic builtins](https://gcc.gnu.org/onlinedocs/gcc-4.1.0/gcc/Atomic-Builtins.html), in particular `__sync_bool_compare_and_swap` and `__sync_val_compare_and_swap`. Unlike `__sync_bool_compare_and_swap`, `__sync_val_compare_and_swap` does not necessarily return a `bool`, so yes, it does seem like something is likely not right here.

Reply:
Thanks. It seems as if the `val` version should return the `type.` Which in this implementation would be an `int` and the value should be the original value of `ptr`. Thanks, nice catch. The threaded Zephyr runtime isn't properly tested. I think I will add a Preprocessor warning about this. Alternatively we could completely disallow it by rejecting LF programs with platform=Zephyr and threading: true